libs/core/synchronization/include/hpx/synchronization/counting_semaphore.hpp

```diff
@@ -342,14 +342,10 @@ namespace hpx {
             typename Mutex = hpx::spinlock>
         class counting_semaphore
         {
-        public:
-            counting_semaphore(counting_semaphore const&) = delete;
-            counting_semaphore& operator=(counting_semaphore const&) = delete;
-            counting_semaphore(counting_semaphore&&) = delete;
-            counting_semaphore& operator=(counting_semaphore&&) = delete;
```

Comment:
Semaphores should be non-movable (see: https://en.cppreference.com/w/cpp/thread/counting_semaphore).

Reply:
Aha, i think I misunderstood something then.

Initially, I thought that the goal was to ref-count copies of the object (similarly to what we do for id_type), and pass around copies of the semaphore object, which would now serve more like an interface to the internal data.
But that confuses copying with referencing, and doesn't conform to the C++ standard either (since we would have to enable copy operations for semaphores). 
I don't really like it.

I now see that, on the barrier example that you had provided, we use the ref-counting on the internal state only for mitigating the issue of the object being destroyed while waiting on a lock. We don't use it for ref-counting copies of the barrier (since the copy/move operations are deleted).
Doing this on the semaphores seems much more reasonable than what I had in mind previously.

Let me know if my new understanding is closer to what you had in mind